Bottom: f9d43b03f5591a81de4ea55530033e28613996ca
Top:    d1967ebf7ced7657fd554041ef0406285bcb2590
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-24 11:49:42 +0800

Refresh of vhost_net-tx-batching

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index df8db5c..0138650 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2589,6 +2589,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	int start;
 	struct vring_used_elem heads[64];
 	struct vring_used_elem __user *used;
+	int i;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
@@ -2600,6 +2601,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	total = vq->avail_idx - vq->last_avail_idx;
 	ret = total = min(total, num);
 
+#if 0
 	while (total) {
 		int ret2 = vhost_get_avail(vq, indices[0],
 			   &vq->avail->ring[last_avail_idx & (vq->num - 1)]);
@@ -2613,6 +2615,18 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		++indices;
 		++last_avail_idx;
 	}
+#endif
+
+	for (i = 0; i < total; i++) {
+		int ret2 = vhost_get_avail(vq, heads[i].id,
+		    &vq->avail->ring[(last_avail_idx + i) & (vq->num - 1)]);
+		if (unlikely(ret2)) {
+			vq_err(vq, "Failed to get descriptors\n");
+			return -EFAULT;
+		}
+		indices[i] = heads[i].id;
+		heads[i].len = 0;
+	}
 
 	start = vq->last_used_idx % (vq->num -1);
 	used = vq->used->ring + start;
